Jenkinsfile

```diff
@@ -1,64 +1,124 @@
-          eval $(perl -I "~/perl5/lib/perl5" -Mlocal::lib)
-          ${K_BIN}/spawn-kserver
-          make -C tests/unit-pass -j$(nproc) os-comparison
-        '''
-      }
-      post {
-        always {
-          archiveArtifacts artifacts: 'tests/unit-pass/*config', allowEmptyArchive: true
+      stages {
+        stage ( 'Set title' ) { steps {
+          script {
+            currentBuild.displayName = "PR ${env.CHANGE_ID}: ${env.CHANGE_TITLE}"
+          }
+        } }
+        stage ( 'Build docker image' ) { steps {
+          script {
+            img = docker.build "${IMG_PR_TAG}"
+          }
+        } }
+        stage ( 'Push to private registry' ) { steps {
+          script {
+            img.push()
+          }
+        } }
+        stage ( 'Checkout source code' ) { steps {
+          script {
+            checkout scm
+          }
+        } }
+        stage ( 'Compile' ) {
+          steps {
+            script { img.inside {
+              sh '''
+                eval $(opam config env)
+                eval $(perl -I "~/perl5/lib/perl5" -Mlocal::lib)
+                export KOMPILE_FLAGS=-O2
+                make -j4 profile-rule-parsing --output-sync=line
+              '''
+            } }
+          }
+          post { success {
+            archiveArtifacts 'dist/timelogs.d/timelogs.csv'
+          } }
+        }
+        stage ( 'Re-compile w/ timeout' ) { steps {
+          timeout(time: 8, unit: 'SECONDS' ) {
+            script { img.inside {
+              sh '''
+                eval $(opam config env)
+                eval $(perl -I "~/perl5/lib/perl5" -Mlocal::lib)
+                make
+              '''
+            } }
+          }
+        } }
+        stage ( 'Test' ) {
+          steps {
+            script { img.inside {
+              sh '''
+                eval $(opam config env)
+                eval $(perl -I "~/perl5/lib/perl5" -Mlocal::lib)
+                ${K_BIN}/spawn-kserver
+                make -C tests/unit-pass -j$(nproc) os-comparison
+              '''
+            } }
+          }
+          post { always {
+            archiveArtifacts artifacts: 'tests/unit-pass/*config', allowEmptyArchive: true
+          } }
         }
+        stage ( 'Test clean target' ) { steps {
+          sh 'make clean'
+          sh '[ $(git clean -xfd 2>&1 | wc -l) -eq 0 ]'
+       } }
+      } // stages of 'Pull Request'
+    }   // Pull Request
+
+    stage ( 'Merged to master' ) {
+      when {
+        branch 'master'
+        beforeAgent true
       }
-    }
-    stage('Test clean target') { steps {
-      sh 'make clean'
-      sh '[ $(git clean -xfd 2>&1 | wc -l) -eq 0 ]'
-    } }
-  }
-}
+      agent any
+      stages {
+        stage ( 'Build docker image' ) { steps {
+          script {
+            img = docker.build "${IMG_RELEASE_TAG}"
+          }
+        } }
+        stage ( 'Push to dockerhub' ) { steps {
+          script {
+            docker.withRegistry( '', 'rvdockerhub' ) {
```

Comment:
The second argument is the credentials id corresponding to the "dockerhub@runtimeverification.com" user.